include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -0,0 +1,402 @@
+                                                         prime_vectors[i]);
+                });
+
+                thread_manager.push_back(std::move(t));
+
+                current_lower_bound = current_upper_bound;
+                current_upper_bound += range_per_thread;
+                primes_in_range = static_cast<Z>(current_upper_bound / std::log(static_cast<double>(current_upper_bound)) -
+                                                 current_lower_bound / std::log(static_cast<double>(current_lower_bound)));
+            }
+
+            prime_vectors.back().reserve(primes_in_range);
+
+            std::thread t([current_lower_bound, upper_bound, &prime_vectors, &pre_generated_primes] {
+                boost::math::detail::segmented_sieve(current_lower_bound, upper_bound, pre_generated_primes, prime_vectors.back());
+            });
+            thread_manager.push_back(std::move(t));
+
+            for(auto &thread : thread_manager)
+            {
+                thread.join();
+            }
+
+            primes.insert(primes.begin(), pre_generated_primes.begin(), pre_generated_primes.end());
+            for(auto &v : prime_vectors)
+            {
+                primes.insert(primes.end(), v.begin(), v.end());
+            }
+        }
+    }
+
+    return std::move(primes.begin(), primes.end(), output);
+}
+
+template<class ExecutionPolicy, class Z, class OutputIterator>
+auto prime_range(ExecutionPolicy&& policy, Z lower_bound, Z upper_bound, OutputIterator output) -> decltype(output)
+{
+    --upper_bound; // Not inclusive, but several methods in boost::math::detail need to be
+    std::vector<Z> primes {};
+    primes.reserve(upper_bound / std::log(static_cast<double>(upper_bound)));
+
+    boost::math::prime_sieve(policy, upper_bound, std::back_inserter(primes));
+
+    auto it{primes.begin()};
+    while(*it < lower_bound && it != primes.end())
+    {
+        ++it;
+    }
+
+    return std::move(it, primes.end(), output);
+}
+#endif //__cplusplus >= 201703
+
+template<class Z, class OutputIterator>
+auto prime_sieve(Z upper_bound, OutputIterator output) -> decltype(output)
+{
+    static_assert(std::is_integral<Z>::value, "No primes for floating point types");
+    BOOST_ASSERT_MSG(upper_bound + 1 < std::numeric_limits<Z>::max(), "Type Overflow");
+
+    --upper_bound; // Not inclusive, but several methods in boost::math::detail need to be
+    std::vector<Z> primes{};
+    primes.reserve(upper_bound / std::log(upper_bound));
+
+    if (upper_bound <= 4096)
+    {
+        boost::math::detail::linear_sieve(upper_bound, primes);
+    }
+
+    else
+    {
+        boost::math::detail::segmented_sieve(static_cast<Z>(2), upper_bound, primes);
+    }
+
+    return std::move(primes.begin(), primes.end(), output);
+}
+
+template<class Z, class OutputIterator>
+auto prime_range(Z lower_bound, Z upper_bound, OutputIterator output) -> decltype(output)
```

Comment:
I'm not sure it's worth including this functionality if it's not computationally more efficient.
After all, this will do the same amount of work, but also allocate extra temporary memory just to omit the primes lower than the lower_bound.  The user can remove the primes less than the lower_bound without using extra temporary memory.

Reply:
My thought process was it would make life easier in that use case. If the user were to remove the primes less than `lower_bound` themselves the memory usage should be the same as moving them the ones they wanted out of the full range [2, upper_bound) they would have received anyway.

Reply:
To be precise, the _peak_ memory allocation will be worse this way. Let's say, 
`x = upper_bound / std::log(static_cast<double>(upper_bound))`, and
`y = (upper_bound - lower_bound) / std::log(static_cast<double>(upper_bound - lower_bound))`
Then using this function, the user allocates size `y` for the final result, and this function allocates `x` for the temporary result of `prime_range(upper)`. So the peak memory usage is `x + y` (ignoring other temporary allocations).

Without this function, the user allocates size `x` for the result, then calls `prime_range(upper)` directly, so the peak memory usage is `x` (again, ignoring other temporary allocations). The user then executes something like `move(begin(result) + lower_bound, end(result), begin(result)); result.resize(upper_bound - lower_bound); result.shrink_to_fit();` to remove the unwanted values and return unused capacity to the system. It's messier and more work for the user, but that's the cost of having lower peak memory overhead.

Is it worth it? Who knows. On reflection, there is no problem with leaving this convenience wrapper in -- users concerned with peak memory allocation more than convenience will just do the second method.